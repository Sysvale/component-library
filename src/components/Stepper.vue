<template>
    <div class="d-flex justify-content-between">
        <div
            v-for="(step, index) in steps"
            :key="index"
            :class="index !== steps.length - 1 ? 'w-100' : ''"
        >
            <div
                class="d-flex align-items-center"
                :id="`step-${index + 1}`"
                @click="!changeStepOnClick || changeStep(index)"
            >
                <div
                    class="d-flex justify-content-center align-items-center cursor-pointer"
                    :class="circleStyle(step)"
                >
                    <span
                        v-if="!step.concluded"
                        class="not-concluded-step-text-color fs-14"
                    >
                        {{ index + 1 }}
                    </span>
                    <check-icon
                        v-else
                        size="1x"
                        class="concluded-step"
                    />
                </div>

                <div 
                    v-if="
                        index !== steps.length - 1 &&
                        steps[index].concluded &&
                        steps[index + 1].concluded
                    "
                    class="concluded-stepper-divider"
                />
                <div 
                    v-else-if="
                        (index !== steps.length - 1 &&
                        index > 0 &&
                        steps[index].concluded &&
                        steps[index - 1].concluded) ||
                        (index !== steps.length - 1 &&
                        index === 0 &&
                        steps[index].concluded)
                    "
                    class="in-progress-stepper-divider"
                />
                <div 
                    v-else-if="index !== steps.length - 1"
                    class="common-stepper-divider"
                />

            </div>
            <div
                class="label-container mt-2"
                :class="!step.active ? 'not-active-step-label' : ''"
            >
                <small>{{ step.label }}</small>
            </div>
        </div>
    </div>
</template>
<script>

import { CheckIcon } from 'vue-feather-icons'

export default {
    props: {
        steps: {
            type: Array,
            default: () => [],
            required: true,
            description: `A object with the properties 'label', and 'concluded'.
                The 'label' property is the text that will be displayed under
                the stepper, and 'concluded' is the status (boolean) of the step.`,
        },
        changeStepOnClick: {
            type: Boolean,
            default: true,
            required: false,
            description: 'Enables or disables the feature to change steps clicking on them.'
        }
    },

    components: {
        CheckIcon,
    },

    methods: {
        circleStyle(step) {
            let style = '';

            if (step.concluded) {
                style += 'concluded-step';
            } 
            
            if (step.active){
                style += 'active-step';
            }

            if (!step.active && !step.concluded) {
                style += ' not-active-step';
            }

            return style;
        },

        changeStep(index) {
            this.steps[index].active = true;
            this.steps.forEach((item, i) => {
                item.active = i === index ? true : false;
            });

            this.$emit('step_changed', index, this.steps[index]);
        }
    }
}
</script>
<style>
    .active-step, .not-active-step, .concluded-step {
        border-radius: 50px;
        min-width: 30px;
        min-height: 30px;
        border: 1.5px;
        border-style: solid;
    }

    .active-step {
        border-color: #00CBAD;
        box-shadow: 0px 0px 8px rgba(67, 228, 204, 0.3);
    }

    .concluded-step { 
        background-color: #00CBAD;
        border-color: #00CBAD;
        width: 30px;
        height: 30px;
    }

    .not-active-step {
        color: #BFC2C5;
        border-color: #BFC2C5;
    }

    .not-concluded-step-text-color {
        color: #00CBAD;
    }

    .common-stepper-divider, .in-progress-stepper-divider, .concluded-stepper-divider {
        height: 3px;
        width: 100%;
    }

    .common-stepper-divider {
        background-color: #BFC2C5;
    }

    .in-progress-stepper-divider {
        background: linear-gradient(90deg, #43E4CC 0%, #BFC2C5 67.57%);
    }

    .concluded-stepper-divider {
        background: #00CBAD;
    }

    .concluded-step {
        color: #fff;
    }

    .label-container {
        max-width: 70px;
        margin-left: -10px;
        line-height: 1.0;
        display: block;
    }

    .fs-14 {
        font-size: 14px;
    }

    .cursor-pointer {
        cursor: pointer;
    }

    .not-active-step-label {
        color: #5d6369;
    }
</style>