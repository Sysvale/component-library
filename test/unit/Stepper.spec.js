import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import Stepper from '../../src/components/Stepper.vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

test('Component is mounted properly', () => {
	let mocked_data = [
		{ label: 'Dummy label 1', concluded: true, active: false },
		{ label: 'Dummy label 3', concluded: false, active: true },
		{ label: 'Dummy label 2', concluded: false, active: false },
	];

	const wrapper = mount(Stepper, {
		localVue,
		propsData: {
			steps: mocked_data,
		},
	});
	expect(wrapper).toMatchSnapshot();
});

describe("Divider styles test", () => {
	test('if all the stepper dividers are rendered as common dividers', () => {
		let mocked_data = [
			{ label: 'Dummy label 1', concluded: false, active: true },
			{ label: 'Dummy label 2', concluded: false, active: false },
			{ label: 'Dummy label 3', concluded: false, active: false },
		];

		const wrapper = mount(Stepper, {
			localVue,
			propsData: {
				steps: mocked_data,
			},
		});
		expect(wrapper.findAll('.common-stepper-divider').length).toBe(2);
	});

	test('if one of the stepper divider is rendered as an in_progress divider and others are rendered as common dividers', () => {
		let mocked_data = [
			{ label: 'Dummy label 1', concluded: true, active: false },
			{ label: 'Dummy label 2', concluded: false, active: true },
			{ label: 'Dummy label 3', concluded: false, active: false },
		];

		const wrapper = mount(Stepper, {
			localVue,
			propsData: {
				steps: mocked_data,
			},
		});
		expect(wrapper.findAll('.in-progress-stepper-divider').length).toBe(1);
		expect(wrapper.findAll('.common-stepper-divider').length).toBe(1);
	});

	test('if one of the stepper divider is rendered as a concluded divider and the other is rendered as an in_progress divider', () => {
		let mocked_data = [
			{ label: 'Dummy label 1', concluded: true, active: false },
			{ label: 'Dummy label 2', concluded: true, active: false },
			{ label: 'Dummy label 3', concluded: false, active: true },
		];

		const wrapper = mount(Stepper, {
			localVue,
			propsData: {
				steps: mocked_data,
			},
		});
		expect(wrapper.findAll('.concluded-stepper-divider').length).toBe(1);
		expect(wrapper.findAll('.in-progress-stepper-divider').length).toBe(1);
	});
});

describe('Change step event tests', () => {
	test('if a event is emited when the stepper is clicked', () => {
		let mocked_data = [
			{ label: 'Dummy label 1', concluded: false, active: true },
			{ label: 'Dummy label 2', concluded: false, active: false },
			{ label: 'Dummy label 3', concluded: false, active: false },
		];

		const wrapper = mount(Stepper, {
			localVue,
			propsData: {
				steps: mocked_data,
			},
		});

		wrapper.find('#step-2').trigger('click');

		expect(wrapper.emitted().step_changed).toBeTruthy();
		expect(wrapper.emitted().step_changed).toEqual([
			[
				1,
				{"active": true, "concluded": false, "label": "Dummy label 2"},
			]
		]);
	});
});
