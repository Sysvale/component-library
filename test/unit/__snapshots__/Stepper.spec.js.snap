// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<div class="d-flex justify-content-between">
  <div class="w-100">
    <div id="step-1" class="d-flex align-items-center">
      <div class="d-flex justify-content-center align-items-center cursor-pointer concluded-step"><svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="concluded-step feather feather-check">
          <polyline points="20 6 9 17 4 12"></polyline>
        </svg></div>
      <div class="in-progress-stepper-divider"></div>
    </div>
    <div class="label-container mt-2 not-active-step-label"><small>Dummy label 1</small></div>
  </div>
  <div class="w-100">
    <div id="step-2" class="d-flex align-items-center">
      <div class="d-flex justify-content-center align-items-center cursor-pointer active-step"><span class="not-concluded-step-text-color fs-14">
                    2
                </span></div>
      <div class="common-stepper-divider"></div>
    </div>
    <div class="label-container mt-2"><small>Dummy label 3</small></div>
  </div>
  <div class="">
    <div id="step-3" class="d-flex align-items-center">
      <div class="d-flex justify-content-center align-items-center cursor-pointer  not-active-step"><span class="not-concluded-step-text-color fs-14">
                    3
                </span></div>
      <!---->
    </div>
    <div class="label-container mt-2 not-active-step-label"><small>Dummy label 2</small></div>
  </div>
</div>
`;
